HPCC-10483 Unicode EditDistanceWithinRadius update

emuharemagic opened the pull request:
Unicode library implementation of EditDistance function was based on two
incorrect assumtions:
1/ The number of Code Units (16-bit words) in a UTF-16 Unicode string is
always the same as the number of Collation Elements associated with this str
2/ Every Code Point always maps into a single Collation Element
Because of that, whenever unicode characters are encoded with multiple
16-bit code units, EditDistance calculation will not be correct.
For example: EditDistance('AVILÉS','AVILAS') will be incorrectly calculated
as 2, instead of 1.
- create unicodeEditDistanceV4() based on the V3, keep all code optimization
  in place but use CharacterIterator instead of RuleBasedCollator
- modify the ulUnicodeLocaleEditDistanceWithinRadius() to use V4 instead of
- update TestEditDistanceWithinRadius module with new tests
  
  ToDo:
    1/ modify ulUnicodeLocaleEditDistance() to use V4 instead of the V2
    2/ Add support for fast, non-locale versions of EditDistance and
    EditDistanceWithinRadius functions; The fast implementations will be based
    on an assumtion that every user perceived unicode character can be
    encoded with no more than 2 16-bit words; They will not use
    the CharacterBreakIterator, and they will outperform the CharacterBreakItera
    based corresponding implementations by an order of magnitude.

Signed-off-by: Edin Muharemagic edin.muharemagic@lexisnexis.com

richardkchapman commented:
@ghalliday Please review

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -278,6 +278,75 @@ inline unsigned char min3(unsigned char a, unsigned char b, unsigned char c)
+        cbi.setText(ustring_);
+        c_offsets_[index] = cbi.first();
+        for (int32_t end = cbi.next(); end != BreakIterator::DONE && length_ < capacity_; end = cbi.next())
+        {
```
Minor: next would be a more common variable name.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -278,6 +278,75 @@ inline unsigned char min3(unsigned char a, unsigned char b, unsigned char c)
+        uint32_t lLen = charLength(index);
+        uint32_t rLen = srcText.charLength(srcIndex);
+        if ( lLen != rLen )
+            return (lLen < rLen) ? -1 : 1;
```
Minor: slightly strange string ordering  'z' < 'aa'. I suspect it makes no difference.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -511,6 +580,121 @@ unsigned unicodeEditDistanceV3(UnicodeString & left, UnicodeString & right, unsi
+
+    UPCList leftCs(bi, left, leftLen);
+    UPCList rightCs(bi, right, rightLen);
+     if (leftCs.isInvalid() || rightCs.isInvalid())
```
Minor: indentation

Commit:
HPCC-10483 Unicode EditDistanceWithinRadius update

Unicode library implementation of EditDistance function was based on two
incorrect assumtions:
1/ The number of Code Units (16-bit words) in a UTF-16 Unicode string is
always the same as the number of Collation Elements associated with this str
2/ Every Code Point always maps into a single Collation Element
Because of that, whenever unicode characters are encoded with multiple
16-bit code units, EditDistance calculation will not be correct.
For example: EditDistance('AVILÉS','AVILAS') will be incorrectly calculated
as 2, instead of 1.

 - create unicodeEditDistanceV4() based on the V3, keep all code optimization
   in place but use CharacterIterator instead of RuleBasedCollator
 - modify the ulUnicodeLocaleEditDistanceWithinRadius() to use V4 instead of
 - update TestEditDistanceWithinRadius module with new tests

 ToDo:
    1/ modify ulUnicodeLocaleEditDistance() to use V4 instead of the V2
    2/ Add support for fast, non-locale versions of EditDistance and
    EditDistanceWithinRadius functions; The fast implementations will be based
    on an assumtion that every user perceived unicode character can be
    encoded with no more than 2 16-bit words; They will not use
    the CharacterBreakIterator, and they will outperform the CharacterBreakItera
    based corresponding implementations by an order of magnitude.


Signed-off-by: Edin Muharemagic <edin.muharemagic@lexisnexis.com>

emuharemagic commented:
@ghalliday I addressed all your comments. Please review the changes

ghalliday commented:
The changes to equals makes the code clearer.
